### ISSUE TITLE
PSTL algos in <numeric> need to move the __init variables

### ISSUE THREAD

**Issue by jwakely** [llvm/llvm-project#118539]
All of `reduce`, `transform_reduce`, `exclusive_scan`, and `inclusive_scan`, `transform_exclusive_scan`, and `transform_inclusive_scan` only have a precondition on the type of `init` that it meets the _Cpp17MoveConstructible_ requirements, so when passing it to the next internal function it needs to be moved, not copied:

```diff
     return transform_inclusive_scan(std::forward<_ExecutionPolicy>(__exec), __first, __last, __result, __binary_op,
-                                    __pstl::__internal::__no_op(), __init);
+                                    __pstl::__internal::__no_op(), std::move(__init));
```

e.g.
https://github.com/llvm/llvm-project/blob/e0ae7793fca0c78919bc41ffd407acb62146ad47/pstl/include/pstl/internal/glue_numeric_impl.h#L33

**Comment by jwakely** [comment]
And they need to move when returning init as well:

```diff
-    return std::make_pair(__result, __init);
+    return std::make_pair(__result, std::move(__init));
```

e.g. https://github.com/llvm/llvm-project/blob/fdd09e9de579e6915d467c1b72efb8c653739602/pstl/include/pstl/internal/numeric_impl.h#L169

**Comment by danhoeflinger** [comment]
@jwakely Thanks for reporting this, and apologies for the delay in our response, it slipped through the cracks.  
We will take a look at this and try to get a fix in to address it.

**Comment by frederick-vs-ja** [comment]
This seems resolved. [libc++'s own implementation](https://github.com/llvm/llvm-project/blob/cad62df49a79df5e5136cfad280c5abc9f62c60b/libcxx/include/__numeric/pstl.h) correctly uses `std::move`, and the old PSTL implementation was removed by #143717.

**Comment by jwakely** [comment]
Yeah let's close it. The problem was fixed downstream in libstdc++ and reported upstream to PSTL.
